lib/projfs.c

```diff
@@ -1451,10 +1505,66 @@ static int check_safe_rel_path(const char *path)
 	return 1;
 }
 
-int projfs_create_proj_dir(struct projfs *fs, const char *path, mode_t mode)
+#define PROJ_XATTR_READ 0x00
+#define PROJ_XATTR_WRITE 0x01
+#define PROJ_XATTR_CREATE 0x02
+
+static char *make_user_xattr_name(const char *segments, unsigned int flags)
 {
-	int fd;
+	char *name;
+
+	name = malloc(PROJ_XATTR_PRE_LEN + strlen(segments) + 1);
+	if (name == NULL)
+		return NULL;
+
+	memcpy(name, PROJ_XATTR_PRE_NAME, PROJ_XATTR_PRE_LEN);
+	strcpy(name + PROJ_XATTR_PRE_LEN, segments);
+
+	if (flags == PROJ_XATTR_READ || check_xattr_reserved(name))
+		return name;
+
+	free(name);
+	errno = EPERM;
+	return NULL;
+}
+
+static int iter_user_xattrs(int fd, struct projfs_attr *attrs,
+			    unsigned int nattrs, unsigned int flags)
+{
+	int set_flags = (flags & PROJ_XATTR_CREATE) ? XATTR_CREATE : 0;
 	int res;
+	unsigned int i;
+
+	if (attrs == NULL)
+		return 0;
+
+	for (i = 0; i < nattrs; i++) {
+		char *name;
+		struct projfs_attr *attr = &attrs[i];
+
+		name = make_user_xattr_name(attr->name, flags);
+		if (name == NULL)
+			return errno;
+
+		if (flags & PROJ_XATTR_WRITE) {
+			res = set_xattr(fd, name, attr->value, &attr->size,
+					set_flags);
+		} else {
+			res = get_xattr(fd, name, attr->value, &attr->size);
+		}
+		if (res == -1)
+			return errno;
```

Comment:
`name` is leaked here.

Reply:
Yes, good catch!  Will fix!!  Thanks for the sharp  👀.